internal/plugin/health_monitor.go

```diff
@@ -0,0 +1,132 @@
+		executorsStore:         executorsStore,
+		sourcesStore:           sourcesStore,
+		pluginRestartStats:     make(map[string]int),
+		healthCheckInterval:    healthCheckInterval,
+	}
+}
+
+func (m *HealthMonitor) Start(ctx context.Context) {
+	go m.monitorSourcePluginHealth(ctx)
+	go m.monitorExecutorPluginHealth(ctx)
+}
+
+func (m *HealthMonitor) monitorSourcePluginHealth(ctx context.Context) {
+	m.log.Info("Starting source plugin supervisor...")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case plugin := <-m.sourceSupervisorChan:
+			m.log.Infof("Restarting source plugin %q...", plugin.name)
+			if source, ok := m.sourcesStore.EnabledPlugins.Get(plugin.name); ok && source.Cleanup != nil {
+				m.log.Debugf("Releasing resources of source plugin %q...", plugin.name)
+				source.Cleanup()
+			}
+
+			// botkube/kubernetes
+			repoPluginPair := fmt.Sprintf("%s/%s", plugin.repo, plugin.name)
+			m.sourcesStore.EnabledPlugins.Delete(repoPluginPair)
+
+			if ok := m.shouldRestartPlugin(repoPluginPair); !ok {
+				m.log.Warnf("Plugin %q has been restarted too many times. Deactivating...", plugin.name)
+				continue
+			}
+
+			p, err := createGRPCClient[source.Source](ctx, m.log, m.logConfig, plugin, TypeSource, m.sourceSupervisorChan, m.healthCheckInterval)
+			if err != nil {
+				m.log.WithError(err).Errorf("Failed to restart plugin %q.", plugin.name)
+				continue
+			}
+
+			m.sourcesStore.EnabledPlugins.Insert(repoPluginPair, p)
+			m.schedulerChan <- repoPluginPair
+		}
+	}
+}
+
+func (m *HealthMonitor) monitorExecutorPluginHealth(ctx context.Context) {
+	m.log.Info("Starting executor plugin supervisor...")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case plugin := <-m.executorSupervisorChan:
+			m.log.Infof("Restarting executor plugin %q...", plugin.name)
+			if executor, ok := m.executorsStore.EnabledPlugins.Get(plugin.name); ok && executor.Cleanup != nil {
+				m.log.Infof("Releasing executors of executor plugin %q...", plugin.name)
+				executor.Cleanup()
+			}
+
+			// botkube/kubectl
+			// TODO: if other naming scheme is used, it might be safer to try guess the name from channel bindings
+			repoPluginPair := fmt.Sprintf("%s/%s", plugin.repo, plugin.name)
+			m.executorsStore.EnabledPlugins.Delete(repoPluginPair)
+
+			if ok := m.shouldRestartPlugin(repoPluginPair); !ok {
+				m.log.Warnf("Plugin %q has been restarted too many times. Deactivating...", plugin.name)
+				continue
+			}
+
+			p, err := createGRPCClient[executor.Executor](ctx, m.log, m.logConfig, plugin, TypeExecutor, m.executorSupervisorChan, m.healthCheckInterval)
+			if err != nil {
+				m.log.WithError(err).Errorf("Failed to restart plugin %q.", plugin.name)
+				continue
+			}
+
+			m.executorsStore.EnabledPlugins.Insert(repoPluginPair, p)
+		}
+	}
+}
+
+func (m *HealthMonitor) shouldRestartPlugin(plugin string) bool {
+	restarts := m.pluginRestartStats[plugin]
+	m.pluginRestartStats[plugin]++
+
+	switch m.policy.Type {
```

Comment:
maybe we can normalize it, to all small letters? so even if I type `restartAgent`  instead of `RestartAgent` it will work.